File: CheatCPP/Source/Utils/TriggerBot.cpp
// C++ / TriggerBot.cpp / Trigger Bot Implementation / Runtime: Unity 2021+
// CoreDev - Automatic Trigger Bot Implementation

#include "Utils/TriggerBot.h"
#include <Windows.h>
#include <vector>
#include <string>

namespace Cheat::Utils
{
    TriggerBot* TriggerBot::s_instance = nullptr;
    
    TriggerBot* TriggerBot::GetInstance()
    {
        if (!s_instance)
        {
            s_instance = new TriggerBot();
        }
        return s_instance;
    }
    
    void TriggerBot::Initialize()
    {
        if (s_instance) return;
        s_instance = new TriggerBot();
    }
    
    void TriggerBot::Shutdown()
    {
        if (s_instance)
        {
            delete s_instance;
            s_instance = nullptr;
        }
    }
    
    TriggerBot::TriggerBot() : 
        m_mainCamera(nullptr),
        m_shouldFire(false),
        m_fireTime(0.0f),
        m_lastFireTime(0.0f)
    {
        // Initialize settings
        m_settings.enabled = false;
        m_settings.triggerKey = VK_LBUTTON;
        m_settings.activationDelay = 0.05f;
        m_settings.randomization = 0.02f;
        m_settings.onlyHeadshots = true;
        m_settings.onlyVisible = true;
        m_settings.maxDistance = 500.0f;
    }
    
    TriggerBot::~TriggerBot() {}
    
    TriggerBotSettings& TriggerBot::GetSettings()
    {
        return m_settings;
    }
    
    void TriggerBot::SetCamera(Camera* camera)
    {
        m_mainCamera = camera;
    }
    
    void TriggerBot::Update()
    {
        if (!m_settings.enabled) return;
        
        // Check if trigger key is pressed
        if (GetAsyncKeyState(m_settings.triggerKey) & 0x8000)
        {
            CheckForTarget();
        }
        
        // Fire if needed
        float currentTime = GetTickCount() / 1000.0f;
        if (m_shouldFire && currentTime >= m_fireTime)
        {
            SimulateFire();
            m_shouldFire = false;
            m_lastFireTime = currentTime;
        }
    }
    
    void TriggerBot::CheckForTarget()
    {
        if (!m_mainCamera) return;
        
        // Placeholder for raycasting
        // In real implementation, we'd raycast from camera center
        
        // Check if hit a valid target
        void* target = nullptr; // Placeholder for hit target
        
        if (IsValidTarget(target))
        {
            // Check if it's a headshot (if enabled)
            if (m_settings.onlyHeadshots && !IsHeadshot(nullptr))
            {
                return;
            }
            
            // Check visibility (if enabled)
            if (m_settings.onlyVisible && !IsVisible(target))
            {
                return;
            }
            
            // Calculate randomized delay
            float delay = m_settings.activationDelay + 
                (static_cast<float>(rand()) / RAND_MAX * 2.0f - 1.0f) * m_settings.randomization;
            m_fireTime = GetTickCount() / 1000.0f + delay;
            m_shouldFire = true;
        }
    }
    
    void TriggerBot::SimulateFire()
    {
        // Placeholder for fire simulation
        // In real implementation, we'd simulate mouse click or call weapon fire
    }
    
    bool TriggerBot::IsValidTarget(void* target)
    {
        if (!target) return false;
        
        // Placeholder - in real implementation, we'd check if it's a player/enemy
        return true;
    }
    
    bool TriggerBot::IsHeadshot(const void* hitInfo)
    {
        // Placeholder - in real implementation, we'd check if hit is on head
        return true;
    }
    
    bool TriggerBot::IsVisible(void* target)
    {
        if (!m_settings.onlyVisible) return true;
        
        // Placeholder - in real implementation, we'd check visibility
        return true;
    }
    
    Transform* TriggerBot::FindDeepChild(Transform* parent, const std::string& name)
    {
        // Placeholder - in real implementation, we'd iterate through children
        return nullptr;
    }
}

File: Cheat/Utils/TriggerBot.cs
// C# / TriggerBot.cs / Trigger Bot System / Runtime: Unity 2021+
// CoreDev - Automatic Trigger Bot with Delay and Randomization

using System;
using UnityEngine;

namespace Cheat.Utils
{
    public class TriggerBot : MonoBehaviour
    {
        // Settings
        public static class Settings
        {
            public static bool Enabled = false;
            public static KeyCode TriggerKey = KeyCode.Mouse0;
            public static float ActivationDelay = 0.05f; // 50ms delay
            public static float Randomization = 0.02f; // 20ms randomization
            public static bool OnlyHeadshots = true;
            public static bool OnlyVisible = true;
            public static float MaxDistance = 500f;
        }
        
        private Camera _mainCamera;
        private bool _shouldFire = false;
        private float _fireTime = 0f;
        private float _lastFireTime = 0f;
        
        private void Awake()
        {
            _mainCamera = Camera.main;
            if (_mainCamera == null)
            {
                Debug.LogError("[CoreDev] TriggerBot: No main camera found!");
                enabled = false;
            }
        }
        
        private void Update()
        {
            if (!Settings.Enabled) return;
            
            // Check if trigger key is pressed
            if (Input.GetKey(Settings.TriggerKey))
            {
                CheckForTarget();
            }
            
            // Fire if needed
            if (_shouldFire && Time.time >= _fireTime)
            {
                SimulateFire();
                _shouldFire = false;
                _lastFireTime = Time.time;
            }
        }
        
        private void CheckForTarget()
        {
            if (_mainCamera == null) return;
            
            // Raycast from camera
            var ray = _mainCamera.ScreenPointToRay(new Vector3(Screen.width / 2f, Screen.height / 2f, 0));
            
            if (Physics.Raycast(ray, out var hit, Settings.MaxDistance))
            {
                var target = hit.collider.gameObject;
                
                // Check if it's a valid target
                if (IsValidTarget(target))
                {
                    // Check if it's a headshot (if enabled)
                    if (Settings.OnlyHeadshots && !IsHeadshot(hit))
                    {
                        return;
                    }
                    
                    // Check visibility (if enabled)
                    if (Settings.OnlyVisible && !IsVisible(target))
                    {
                        return;
                    }
                    
                    // Calculate randomized delay
                    var delay = Settings.ActivationDelay + UnityEngine.Random.Range(-Settings.Randomization, Settings.Randomization);
                    _fireTime = Time.time + delay;
                    _shouldFire = true;
                }
            }
        }
        
        private bool IsValidTarget(GameObject target)
        {
            if (target == null) return false;
            
            // Check if it's a player
            var player = target.GetComponentInParent<MonoBehaviour>();
            if (player != null)
            {
                var typeName = player.GetType().Name;
                if (typeName.Contains("Player") || typeName.Contains("Character"))
                {
                    return true;
                }
            }
            
            // Check if it's an enemy
            var enemy = target.GetComponentInParent<MonoBehaviour>();
            if (enemy != null)
            {
                var typeName = enemy.GetType().Name;
                if (typeName.Contains("Enemy") || typeName.Contains("Zombie"))
                {
                    return true;
                }
            }
            
            return false;
        }
        
        private bool IsHeadshot(RaycastHit hit)
        {
            var target = hit.collider.gameObject;
            var head = FindDeepChild(target.transform, "Head");
            
            if (head != null)
            {
                // Check if hit point is near head
                var distance = Vector3.Distance(hit.point, head.position);
                return distance < 0.3f; // 30cm tolerance
            }
            
            // Check if hit collider is head collider
            if (hit.collider.name.Contains("Head", StringComparison.OrdinalIgnoreCase))
            {
                return true;
            }
            
            return false;
        }
        
        private Transform FindDeepChild(Transform parent, string name)
        {
            foreach (Transform child in parent.GetComponentsInChildren<Transform>(true))
            {
                if (child.name.Contains(name, StringComparison.OrdinalIgnoreCase))
                {
                    return child;
                }
            }
            return null;
        }
        
        private bool IsVisible(GameObject target)
        {
            if (!Settings.OnlyVisible) return true;
            
            var renderer = target.GetComponentInChildren<Renderer>();
            if (renderer == null) return true;
            
            // Simple visibility check
            return renderer.isVisible;
        }
        
        private void SimulateFire()
        {
            // Simulate left mouse button press and release
            Debug.Log("[CoreDev] TriggerBot: Firing!");
            
            // In actual implementation, we'd use:
            // 1. InputSimulator to simulate mouse clicks
            // 2. Or directly call weapon fire methods
            // 3. Or use Unity's Input system
            
            // This is a placeholder for the actual implementation
            var fireEvent = new InputEvent
            {
                type = InputEventType.MouseButtonDown,
                button = 0
            };
            
            // Process fire event
            ProcessInputEvent(fireEvent);
            
            // Release after a frame
            var releaseEvent = new InputEvent
            {
                type = InputEventType.MouseButtonUp,
                button = 0
            };
            
            ProcessInputEvent(releaseEvent);
        }
        
        private void ProcessInputEvent(InputEvent evnt)
        {
            // This would be handled by the input system
            // In a real implementation, we'd inject the event into Unity's input system
        }
        
        // Input event struct
        private struct InputEvent
        {
            public InputEventType type;
            public int button;
        }
        
        private enum InputEventType
        {
            MouseButtonDown,
            MouseButtonUp,
            KeyDown,
            KeyUp
        }
    }
}
